docs/source/datasets.rst

```diff
@@ -3,6 +3,51 @@ torchtext.datasets
 
 .. currentmodule:: torchtext.datasets
 
+
+.. _datapipes_warnings:
+
+.. warning::
+
+    The datasets supported by torchtext are datapipes from the `torchdata
+    project <https://pytorch.org/data/beta/index.html>`_, which is still in Beta
+    status. This means that the API is subject to change without deprecation
+    cycles. In particular, we expect a lot of the current idioms to change with
+    the eventual release of ``DataLoaderV2`` from ``torchdata``.
+
+    Here are a few recommendations regarding the use of datapipes:
+
+    - For shuffling the datapipe, do that in the DataLoader: ``DataLoader(dp, shuffle=True)``.
+      You do not need to call ``dp.shuffle()``, because ``torchtext`` has
+      already done that for you. Note however that the datapipe won't be
+      shuffled unless you explicitly pass ``shuffle=True`` to the DataLoader.
+
+    - When using multi-processing (``num_workers=N``), use the builtin ``worker_init_fn``::
+
+            from torch.utils.data.backward_compatibility import worker_init_fn
+            DataLoader(dp, num_workers=4, worker_init_fn=worker_init_fn, drop_last=True)
+
+      This will ensure that data isn't duplicated across workers.
+
+    - We also recommend using ``drop_last=True``. Without this, the batch sizes
+      at the end of an epoch may be very small in some cases (smaller than with
+      other map-style datasets). This might affect accuracy greatly especially
+      when batch-norm is used. ``drop_last=True`` ensures that all batch sizes
+      are equal.
+
+    - Distributed training with ``DistributedDataParallel`` is not yet entirely
+      stable / supported, and we don't recommend it at this point. It will be
+      better supported in DataLoaderV2. If you still wish to use DDP, make sure
+      that:
+
+      - All workers (DDP workers *and* DataLoader workers) see a different part
+        of the data. You might need to call ``dp = dp.apply_sharding(world_size,
+        rank)``.
```

Comment:
@parmeet good catch - users would need to wrap the dp again before calling `apply_sharding()`. I'll update the comment

```py
from torchdata.datapipes.iter import IterableWrapper, ShardingFilter

a = IterableWrapper(range(30))

a = ShardingFilter(a)
a.apply_sharding(num_of_instances=2, instance_id=0)

a = ShardingFilter(a)  # without this, only the second sharding (5) applies
a.apply_sharding(num_of_instances=5, instance_id=0)

list(a)
# [0, 10, 20]
```

Reply:
I think this sounds neat. This would also avoid applying sharding filter multiple times. Given that datasets already wrapped the dp inside sharding filter, all we need to do is provide the implementation of worker_init_fn that does sharding according to world size and worker IDs. 

```python
def worker_init_fn(worker_id, world_size, rank):
    info = torch.utils.data.get_worker_info()
    num_workers = info.num_workers
    datapipe = info.dataset
    datapipe.apply_sharding(world_size * worker_num, rank * worker_num + worker_id)

# Initialize DataLoader within each rank with above function
dl = DataLoader(dataset, worker_init_fn=functools.partial(worker_init_fn, world_size=world_size, rank=rank), ...)
```

Does this sounds about right @ejguan?

Reply:
Probably best to get work_size and rank from within the function? One can use `int(os.environ["RANK"])` and `int(os.environ["WORLD_SIZE"])`. Also for clarity maybe we could rename these to `num_ddp_workers`, `num_dataloader_workers`, `ddp_worker_id`, `dataloader_worker_id`.

It might be worth documenting both ways? The second one is somehow more complex, because it exposes the DataLoader sharding level.

Reply:
@parmeet That is the solution in my mind.

@NicolasHug That would be the ideal case. I believe worker processes inherit the environment variable like `RANK` and `WORLD_SIZE`. But, considering that I only tested it in multiprocessing environment and my bare knowledge on distributed processes, I was hesitating to recommend to fetch such env in worker processes.

Reply:
SGTM! I think we can leave the details of implementation to user. At higher level, probably it is sufficient to communicate the workaround with worker_init_fn :)

Reply:
Please do not use ENV variables. There are built-in things for it:
```python
import torch.distributed as dist
if dist.is_available():
        total_workers *= dist.get_world_size()
        global_worker_id = dist.get_rank()*info.num_workers + global_worker_id
```

I will submit PR(s) to apply it automatically this week.

The only remaining problem is the possibility of having a different number of processes on different ranks. I prefer to keep one `.sharding_filter` for it, but perhaps would have to introduce sharing ranks/levels into it.

Reply:
Thanks @VitalyFedyunin for the feedback. Just so I understand, do you mean to submit PR for updating [worker_init_fn](https://github.com/pytorch/pytorch/blob/d90012689f5b99bbc289da598ad982f829687288/torch/utils/data/backward_compatibility.py#L4) such that it not only works for multi-processing case (as of now) but shall also take into account distributed settings like the one propose here https://github.com/pytorch/text/pull/1755#discussion_r886909628?

Reply:
Yes.  And also classical DataLoader will automatically add this worker_init_fn pieces to all DataPipe-based Datasets. https://github.com/pytorch/pytorch/pull/78631